examples/3.routes/8.tolls-and-ferries/interface.js

```diff
@@ -0,0 +1,241 @@
+import { getDurationString } from '../../../utils';
+
+// Array used for journey overview.
+let journeyLegs = [];
+
+/**
+ * Render journey overview.
+ *
+ * @param data {object} route specification
+ */
+export const renderRouteData = data => {
+  // the total distance of the route, in meters converted to km
+  const routeDistance = data.distance ? `${(data.distance / 1000).toFixed(0)} km` : 'Unknown';
+
+  // the total energy used of the route, in kWh
+  const routeEnergy = data.consumption ? `${data.consumption.toFixed(2)} kWh` : 'Unknown';
+
+  // the total duration of the journey (including charge time), in seconds
+  document.getElementById('duration').innerHTML = `${getDurationString(data.duration ?? 0)}`;
+
+  let extraFerryStops = 0;
+  let origin = '';
+  let destination = '';
+  // Convert the legs we receive from the API to journey overview legs we can display.
+  // First check if the route has loaded yet
+  if (data) {
+    console.log(data);
+    // Determine origin and destination
+    origin = data.legs[0].origin.properties.name;
+    destination = data.legs[data.legs.length - 1].destination.properties.name;
+    // Go through the legs we receive from the backend - stations and final.
+    for (let i = 0; i < data.legs.length; i++) {
```

Comment:
Instead of this `for` I would do:

```ts
  var journeyLegs = data.legs.reduce((legAcc, leg) => {
    // Converting step list into Legs using FE rules
    var stepLegs = leg.steps.reduce((stepAcc, step) => {
      switch (step.type) {
        case 'ferry': {
          if (stepAcc[stepAcc.length - 1].type !== 'ferry') {
            stepAcc.push({
              type: 'ferry',
              distance: 0,
              duration: 0,
              tollsOnLeg: false,
            });
          }

          break;
        }
        default: {
          if (stepAcc.length === 0 || stepAcc[stepAcc.length - 1].type === 'ferry') {
            stepAcc.push({
              type: 'road',
              distance: 0,
              duration: 0,
              tollsOnLeg: false,
            });
          }

          break;
        }
      }

      stepAcc[stepAcc.length - 1].distance += step.distance;
      stepAcc[stepAcc.length - 1].duration += step.duration;
      stepAcc[stepAcc.length - 1].tollsOnLeg = stepAcc[stepAcc.length - 1].tollsOnLeg || step.type === 'toll';

      return stepAcc;
    }, []);
    stepLegs[0].origin = leg.origin;
    stepLegs[0].rangeStartPercentage = leg.rangeStartPercentage;
    stepLegs[stepLegs.length - 1].destination = leg.destination;
    stepLegs[stepLegs.length - 1].rangeEndPercentage = leg.rangeEndPercentage;

    // Adding the charging station or via as a Leg
    if (leg.type !== 'final') {
      stepLegs.push(leg);
    }

    legAcc.push(...stepLegs);

    return legAcc;
  }, []);
```

Method `splitLeg` seems to be over complicated.

Reply:
I might be missing context here, but @cosmin-petrescu that is indeed somewhat similar to what we did for Discover. However, we couldn't know in advance which [`StepType`](https://github.com/chargetrip/types/blob/503343fa64c8e94efbdcdecff9bad929d4408bfe/src/graphql.ts#L4167-L4181) to _add_ and which to _ignore_ if the `enum` were to be updated.

 For example, `StepType.WALKING` was not supposed to be included for Discover, but the `StepType.CROSSBORDER` step was.
 
Would the above not also push and add the `WALKING` step distance and duration as of type `ROAD` and need a possible extra case?

For what it's worth, I ended up using a similar `for...of` myself to make it a bit easier to understand for people compared to `.reduce` / `.fold` methods in languages, but that doesn't change how the code works.

```ts
const legSteps = [];

for (const step of leg.steps) {
  // This might return undefined and throw an error on accessing `type` below?
  const previousStep = legSteps[legSteps.length - 1];

  switch (step.type) {
    case StepType.FERRY:
      if (previousStep.type !== StepType.FERRY) {
        stepLegs.push({
          type: StepType.FERRY,
          distance: step.distance,
          duration: step.duration,
          tollsOnLeg: false,
        });
      }
      
      break;
    default:
      if (legSteps.length === 0 || legSteps.type === StepType.FERRY) {
        stepLegs.push({
          type: StepType.ROAD,
          distance: step.distance,
          duration: step.duration,
          tollsOnLeg: step.type === StepType.TOLL,
        });
      }

      break;
  }
}

// ... modify legs
```

I would otherwise perhaps rename `stepLegs` to `legSteps`, if they are the "steps" of a "leg" similar to how the "legs" of a "journey" results in the name `journeyLegs`?

Reply:
Depends on the context you want to apply, for this example we took into account only the `ferry`, but if you want to render differently the `cross_border` or `walking`, you need to add cases for them.
Atm, this is only extracting the `ferry` into a separated leg, or better say convert the response structure to a structure you need to render on FE.